### DISCUSSION TITLE
Support for an extensive URI scheme

### DISCUSSION THREAD

**Discussion by lilyremigia** [ppy/osu#14231] [Ideas]
With how flexible the osu!lazer client is, and how big the community is, it makes sense to support an extensive URI scheme.
My recommendation is as follows:

### osu://import
Which can download and import any kind of user-made content that may be shared. Such as:
- `osu://import/beatmapset/id/163112` which would download - of course - [this](https://osu.ppy.sh/beatmapsets/163112/).
- `osu://import/beatmapset/file/https://link-to-osz-or-zip-file.osz`.

Other types would be skins, rulesets, replay/scores(?), collections:
- `osu://import/skin/file/http://you-get-the-idea.osk-or.zip`
- `osu://import/ruleset/file/http://you-get-the-idea.zip-there-could-be-an-extension-for-this-too-but-that-is-a-different-issue`
- `osu://import//replay/file/http://you-get-the-idea.osr`
- `osu://import/collection/file/http://idk-what-the-extension-for-this-is-i-was-just-looking-at-the-settings-for-extra-ideas`

### osu://overlay
Opens up a specific overlay, or a part in the overlay. Would be helpful in helping users in chat, or on a wiki. Only for those that don't have a web equivalent.
Examples:
- `osu://overlay/settings/keybinds/osu` would open up [keybind settings for the osu standard ruleset](https://user-images.githubusercontent.com/13798839/129072598-0fd07405-1c10-4e33-8239-7cd4e6464721.jpg).
- `osu://overlay/play/solo` could be used as a quick shortcut
- `osu://overlay/edit/id/163112` and `osu://overlay/edit/file/http://link-to-.osz` could be helpful to modders, I think.
- `osu://overlay/nowplaying`, `osu://overlay/volume`, etc. I don't feel like is required in the scheme, but could be used privately in-app for helping users in chat.

### osu://multi
- `osu://multi/rt/00` would obviously join a multiplayer (Real-Time) game, as opposed to
- `osu://multi/pl/00` which would join a playlist/timeshift, since I'm pretty sure those two are handled separately.
Also a post argument of `?pwd=1234` could be added, so users when inviting don't get prompted for a password.

Of course, the team is welcome to tweaks or optimize this as it fits. A full list of these when implemented would also be a welcome addition to the wiki.

**Comment by lilyremigia** [comment]
To the overlay part, the ones that have a web equivalent: It might be a better idea to not add to the "scheme", and just open in-client, when clicked in-game.

**Comment by bdach** [comment]
Imports from online sources should be handled by web browser via the MIME type. osu! is not a browser/file downloader, and embedding URIs into URIs is an overreach.

And yes, all the web views will open in-client rather than externally once they are in place.

**Comment by LittleEndu** [comment]
Wouldn't this have to overwrite stable being the default handler of `osu://` scheme? It's currently used by stable to download beatmaps via osu direct (`osu://b/2397741`) and link to beatmap hitobjects in currently opened beatmap (`osu://edit/00:09:734 (6)`).

**Reply by bdach** [reply]
...None of the above links are conflicting with the original proposal? And yes, if you are talking about the OS level of things, there can ever only be one default handler of a URI scheme, can there?

**Comment by peppy** [comment]
most of the proposed are already implemented or present on stable and will be once the features come online.
